File: packages/shared-types/src/narration.ts
/**
 * Trigger that caused a narration to fire.
 */
export type NarrationTrigger =
  | "approaching"   // Geofence entry — vehicle nearing a tour stop
  | "arrived"        // At the stop location
  | "departed"       // Left the stop location
  | "manual";        // User-initiated narration request

/**
 * Payload delivered to the TTS engine for a narration event.
 * Contains everything needed to generate a spoken summary for a tour stop.
 */
export interface NarrationPayload {
  /** Tour stop that triggered this narration */
  tourStopId: string;
  /** Listing at this stop */
  listingId: string;
  /** What triggered the narration */
  trigger: NarrationTrigger;
  /** Pre-composed narration text (ready for TTS) */
  narrationText: string;
  /** Structured listing summary for display alongside narration */
  listingSummary?: {
    address: string;
    price: string;
    beds: number | null;
    baths: number | null;
    sqft: number | null;
    highlights?: string[];
  };
  /** Navigation context — where the user is relative to the stop */
  navigationContext?: {
    distanceMeters: number;
    durationSeconds: number;
    /** "next" = upcoming stop, "current" = at stop now */
    relation: "next" | "current";
  };
}

/**
 * Type of proximity event detected by the geofence system.
 */
export type ProximityEventType = "approaching" | "arrived" | "departed";

/**
 * A proximity event fired when the device enters, reaches, or leaves
 * a geofenced area around a tour stop.
 */
export interface ProximityEvent {
  /** Tour this event belongs to */
  tourId: string;
  /** Tour stop that triggered the event */
  tourStopId: string;
  /** Listing at this stop */
  listingId: string;
  /** Type of proximity change */
  type: ProximityEventType;
  /** Device location when the event fired */
  location: {
    lat: number;
    lng: number;
  };
  /** Distance from the stop in meters */
  distanceMeters: number;
  /** ISO 8601 timestamp */
  timestamp: string;
}

File: apps/mobile/lib/models/narration.dart
import 'package:json_annotation/json_annotation.dart';

part 'narration.g.dart';

/// Mirrors NarrationPayload from packages/shared-types/src/narration.ts.
@JsonSerializable()
class NarrationPayload {
  final String tourStopId;
  final String listingId;
  final String trigger;
  final String narrationText;
  final NarrationListingSummary? listingSummary;
  final NarrationNavigationContext? navigationContext;

  const NarrationPayload({
    required this.tourStopId,
    required this.listingId,
    required this.trigger,
    required this.narrationText,
    this.listingSummary,
    this.navigationContext,
  });

  factory NarrationPayload.fromJson(Map<String, dynamic> json) =>
      _$NarrationPayloadFromJson(json);

  Map<String, dynamic> toJson() => _$NarrationPayloadToJson(this);
}

@JsonSerializable()
class NarrationListingSummary {
  final String address;
  final String price;
  final int? beds;
  final int? baths;
  final int? sqft;
  final List<String>? highlights;

  const NarrationListingSummary({
    required this.address,
    required this.price,
    this.beds,
    this.baths,
    this.sqft,
    this.highlights,
  });

  factory NarrationListingSummary.fromJson(Map<String, dynamic> json) =>
      _$NarrationListingSummaryFromJson(json);

  Map<String, dynamic> toJson() => _$NarrationListingSummaryToJson(this);
}

@JsonSerializable()
class NarrationNavigationContext {
  final int distanceMeters;
  final int durationSeconds;
  final String relation;

  const NarrationNavigationContext({
    required this.distanceMeters,
    required this.durationSeconds,
    required this.relation,
  });

  factory NarrationNavigationContext.fromJson(Map<String, dynamic> json) =>
      _$NarrationNavigationContextFromJson(json);

  Map<String, dynamic> toJson() => _$NarrationNavigationContextToJson(this);
}

/// Mirrors ProximityEvent from packages/shared-types/src/narration.ts.
@JsonSerializable()
class ProximityEvent {
  final String tourId;
  final String tourStopId;
  final String listingId;
  final String type;
  final ProximityLocation location;
  final int distanceMeters;
  final String timestamp;

  const ProximityEvent({
    required this.tourId,
    required this.tourStopId,
    required this.listingId,
    required this.type,
    required this.location,
    required this.distanceMeters,
    required this.timestamp,
  });

  factory ProximityEvent.fromJson(Map<String, dynamic> json) =>
      _$ProximityEventFromJson(json);

  Map<String, dynamic> toJson() => _$ProximityEventToJson(this);
}

@JsonSerializable()
class ProximityLocation {
  final double lat;
  final double lng;

  const ProximityLocation({required this.lat, required this.lng});

  factory ProximityLocation.fromJson(Map<String, dynamic> json) =>
      _$ProximityLocationFromJson(json);

  Map<String, dynamic> toJson() => _$ProximityLocationToJson(this);
}
